prometheus/doc.go

```diff
@@ -61,51 +62,135 @@
+// NewConstSummary (and their respective Must… versions). That will happen in
+// the Collect method, while the Describe method has to return separate Desc
+// instances, the very same you later use to create the constant “throw-away”
+// metrics. NewDesc comes in handy to create those Desc instances.
+//
+// The Collector example illustrates the use-case. You can also look at the
+// source code of the processCollector (mirroring process metrics), the
+// goCollector (mirroring Go metrics), or the exvarCollector (mirroring expvar
+// metrics) as examples that are used in this package itself.
+//
+// If you just need to call a function to get a single float value to collect as
+// a metric, GaugeFunc, CounterFunc, or UntypedFunc might be interesting
+// shortcuts.
+//
+// Advanced Uses of the Registry
+//
+// While MustRegister is the by far most common way of registering a Collector,
+// sometimes you might want to handle the errors the registration might
+// cause. As suggested by the name, MustRegister panics if an error occurs. With
+// the Register function, the error is returned and can be handled.
+//
+// An error is returned if the registered Collector is incompatible or
+// inconsistent with already registered metrics. The registry aims for
+// consistency of the collected metrics according to the Prometheus data
+// model. Inconsistencies are ideally detected at registration time, not at
+// collect time. The former will usually be detected at start-up time of a
+// program, while the latter will only happen at scrape time, possibly not even
+// on the first scrape if the inconsistency only becomes relevant later. That is
+// the main reason why a Collector and a Metric have to describe themselves to
+// the registry.
+//
+// A particular error case to handle is if a Collector has already be registered
+// (or a different Collector that collects the same metrics). Registration fails
+// in that case, but you can detect from the kind of error what has happened. A
+// common pattern is that you want to register the new Collector if an equal
+// Collector hasn't been registered before, but otherwise you want to switch
+// over to the previously registered Collector. This pattern is supported by the
+// RegisterOrGet function.
+//
+// So far, everything we did operated on the so-called default registry, as it
+// can be found in the global DefaultRegistry variable. With NewRegistry, you
+// can create a custom registry, or you can even implement the Registry
+// interface yourself. The methods Register and Unregister work in the same way
+// on a custom registry as the global functions Register and Unregister on the
+// default registry. The other global functions are not methods in the Registry
+// interface but there are utility functions that implement the same
+// functionality on top of the interface and take a Registry instance as a
+// parameter. Those are: MustRegister → MustRegisterWith, RegisterOrGet →
```

Comment:
Having an interface is the "fail safe" approach. Because this library is used quite widely, I assume a lot of requirements. I have observed the recent tendency to go for interfaces in doubt, and may people complaining about std library types that were not done as interfaces.

The normal use case is anyway to use the default registry. So only power user are even concerned with the registry as an own type. So I expect a higher percentage of "weird" use cases there, which we cannot anticipate right now. In any case, we don't have to optimize the registry itself for the naive user, as the latter will use the top-level functions.